cpg_workflows/large_cohort/combiner.py

```diff
@@ -9,23 +9,123 @@
+    j.image(config_retrieve(['workflow', 'driver_image']))
+    j.memory(config_retrieve(['combiner', 'memory']))
+    j.storage(config_retrieve(['combiner', 'storage']))
+
+    # set this job to be non-spot (i.e. non-preemptible)
+    # previous issues with preemptible VMs led to multiple simultaneous QOB groups processing the same data
+    j.spot(config_retrieve(['combiner', 'preemptible_vms'], False))
+    return j
+
+
+def get_vds_ids_output(vds_id: int) -> tuple[str, list[str]]:
+    get_vds_analysis_query: DocumentNode = gql(
+        """
+            query getVDSByAnalysisIds($vds_id: Int!) {
+                analyses(id: {eq: $vds_id}) {
+                    output
+                    sequencingGroups {
+                        id
+                    }
+                }
+            }
+        """,
+    )
+    query_results: dict[str, Any] = query(get_vds_analysis_query, variables={'vds_id': vds_id})
+    vds_path: str = query_results['analyses'][0]['output']
+    vds_sgids: list[str] = [sg['id'] for sg in query_results['analyses'][0]['sequencingGroups']]
+    return (vds_path, vds_sgids)
+
+
+def combiner(cohort: Cohort, output_vds_path: str, save_path: str) -> PythonJob:
+    workflow_config = config_retrieve('workflow')
+    combiner_config = config_retrieve('combiner')
+
+    tmp_prefix_for_withdrawals: str = slugify(
+        f'{cohort.analysis_dataset.tmp_prefix()}{cohort.name}-{workflow_config["sequencing_type"]}-{combiner_config["vds_version"]}',
+    )
+
+    vds_paths: list[str] = []
+    sg_ids_in_vds: list[str] = []
+    sgs_for_withdrawal: list[str] = []
+    new_sg_gvcfs: list[str] = []
+    cohort_sgs: list[SequencingGroup] = cohort.get_sequencing_groups(only_active=True)
+    cohort_sg_ids: list[str] = cohort.get_sequencing_group_ids(only_active=True)
+    gvcf_external_header: str | None = None
+    sequencing_group_names: list[str] | None = None
+
+    if combiner_config.get('vds_analysis_ids', None) is not None:
+        for vds_id in combiner_config['vds_analysis_ids']:
+            tmp_query_res, tmp_sg_ids_in_vds = get_vds_ids_output(vds_id)
+            vds_paths.append(tmp_query_res)
+            sg_ids_in_vds = sg_ids_in_vds + tmp_sg_ids_in_vds
+            sgs_for_withdrawal = [sg for sg in sg_ids_in_vds if sg not in cohort_sg_ids]
+
+    if combiner_config.get('merge_only_vds', False) is not True:
+        # Get SG IDs from the cohort object itself, rather than call Metamist.
+        # Get VDS IDs first and filter out from this list
+        new_sg_gvcfs = [str(sg.gvcf) for sg in cohort_sgs if sg.gvcf is not None and sg.id not in sg_ids_in_vds]
+        if new_sg_gvcfs:
+            gvcf_external_header = new_sg_gvcfs[0]
+            sequencing_group_names = [
+                str(sg.id) for sg in cohort_sgs if sg.gvcf is not None and sg.id not in sg_ids_in_vds
+            ]
+
+    # if new_sg_gvcfs and len(new_sg_gvcfs) == 0 and len(vds_paths) <= 1:
+    #     raise Exception
```

Comment:
```suggestion
    if not (new_sg_gvcfs or vds_paths):
        raise Exception
```

This still feels like a valid exception, otherwise you continue to the combiner but it won't be doing any combining?

Reply:
Yeah so this will actually throw an exception if you want to only remove samples. Let me update this to make sure it checks that there are no SGs to remove as well.